src/ring_theory/tensor_product.lean

```diff
@@ -1,43 +1,307 @@
 /-
 Copyright (c) 2020 Scott Morrison. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Scott Morrison
+Authors: Scott Morrison, Johan Commelin
 -/
 
 import linear_algebra.tensor_product
-import algebra.algebra.basic
+import algebra.algebra.tower
 
 /-!
 # The tensor product of R-algebras
 
-We construct the R-algebra structure on `A ⊗[R] B`, when `A` and `B` are both `R`-algebras,
-and provide the structure isomorphisms
+Let `R` be a (semi)ring and `A` an `R`-algebra.
+In this file we:
 
-* `R ⊗[R] A ≃ₐ[R] A`
-* `A ⊗[R] R ≃ₐ[R] A`
-* `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+- Define the `A`-module structure on `A ⊗ M`, for an `R`-module `M`.
+- Define the `R`-algebra structure on `A ⊗ B`, for another `R`-algebra `B`.
+  and provide the structure isomorphisms
+  * `R ⊗[R] A ≃ₐ[R] A`
+  * `A ⊗[R] R ≃ₐ[R] A`
+  * `A ⊗[R] B ≃ₐ[R] B ⊗[R] A`
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+
+## Main declaration
+
+- `linear_map.base_change A f` is the `A`-linear map `A ⊗ f`, for an `R`-linear map `f`.
 
-The code for
-* `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
-is written and compiles, but takes longer than the `-T100000` time limit,
-so is currently commented out.
 -/
 
 universes u v₁ v₂ v₃ v₄
 
-namespace algebra
-
 open_locale tensor_product
 open tensor_product
 
 namespace tensor_product
 
+variables {R A M N P : Type*}
+
+section restrict_scalars
+variables [comm_semiring R] [comm_semiring A] [algebra R A]
+variables [add_comm_monoid M] [module A M]
+variables [add_comm_monoid N] [module A N]
+
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.module : module R (M ⊗[A] N) :=
+show module R (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.module R A (M ⊗ N)
```

Comment:
Reiterating [this hidden comment](https://github.com/leanprover-community/mathlib/pull/4773#discussion_r550436289): I think this instance already exists under `[is_scalar_tower R A M] [is_scalar_tower R A N]` assumptions, which probably means this instance is a bad idea.

Reply:
Noted, and this also looks suspicious to me. Looking at `linear_map.restrict_scalars`, I wonder do we need it?

Reply:
I have just added https://github.com/leanprover-community/mathlib/pull/4773/commits/89b22a27cddd615c21ef7fc228c1c5f55143e735 which removes this instance.